mmdet/datasets/samplers/infinite_sampler.py

```diff
@@ -0,0 +1,86 @@
+import itertools
+
+import numpy as np
+import torch
+from mmcv.runner import get_dist_info
+from torch.utils.data.sampler import Sampler
+
+
+class DistributedInfiniteGroupSampler(Sampler):
+    """Similar to `DistributedGroupSampler` but designed for `IterationBased`
+    runner.
+
+    The implementation logic is referred to
+    https://github.com/facebookresearch/detectron2/blob/main/detectron2/data/samplers/grouped_batch_sampler.py
+
+    Args:
+        dataset (object): The dataset.
+        samples_per_gpu (int): Number of training samples on each GPU.
+            Default: 1.
+        num_replicas (int): Number of processes participating in
+            distributed training.
+        rank (int): Rank of current process. Default: None.
+        seed (int): Random seed. Default: 0.
+        shuffle (bool): Whether shuffle the dataset or not. Default: True.
+    """  # noqa: W605
+
+    def __init__(self,
+                 dataset,
+                 samples_per_gpu=1,
+                 num_replicas=None,
+                 rank=None,
+                 seed=0,
+                 shuffle=True):
+        _rank, _num_replicas = get_dist_info()
+        if num_replicas is None:
+            num_replicas = _num_replicas
+        if rank is None:
+            rank = _rank
+        self.rank = rank
+        self.num_replicas = num_replicas
+        self.dataset = dataset
+        self.samples_per_gpu = samples_per_gpu
+        self.seed = seed if seed is not None else 0
+        self.shuffle = shuffle
+
+        assert hasattr(self.dataset, 'flag')
+        self.flag = self.dataset.flag
+        self.group_sizes = np.bincount(self.flag)
+        # buffer used to save indices of each group
+        self.buffer_per_group = {k: [] for k in range(len(self.group_sizes))}
+
+        self.size = len(dataset)
+        self.indices = self._indices_of_rank()
+
+    def _infinite_indices(self):
+        """Infinitely yield a sequence of indices."""
+        g = torch.Generator()
+        g.manual_seed(self.seed)
+        while True:
+            if self.shuffle:
+                yield from torch.randperm(self.size, generator=g).tolist()
+            else:
+                yield from torch.arange(self.size).tolist()
+
+    def _indices_of_rank(self):
+        """Slice the infinite indices by rank."""
+        yield from itertools.islice(self._infinite_indices(), self.rank, None,
+                                    self.num_replicas)
+
+    def __iter__(self):
+        # once batch size is reached, yield the indices
+        for idx in self.indices:
+            flag = self.flag[idx]
+            group_buffer = self.buffer_per_group[flag]
+            group_buffer.append(idx)
+            if len(group_buffer) == self.samples_per_gpu:
+                yield group_buffer[:]
+                del group_buffer[:]
+
+    def __len__(self):
+        """Length of base dataset."""
+        return self.size
```

Comment:
We need to discuss this dummy length.  This dummy length may be used to check whether is at the end of an `epoch`. Although there seems to be no error,  It is not reasonable when is compared to the length of `DistributedGroupSampler`. There may be other potential problems due to our `IterBasedRunner` using an epoch based dataloader to Imitate the behavior.